.circleci/regenerate.py

```diff
@@ -1,13 +1,121 @@
-#!/usr/bin/env python
+#!/usr/bin/env python3
+
+"""
+This script should use a very simple, functional programming style.
+Avoid Jinja macros in favor of native Python functions.
+
+Don't go overboard on code generation; use Python only to generate
+content that can't be easily declared statically using CircleCI's YAML API.
+
+Data declarations (e.g. the nested loops for defining the configuration matrix)
+should be at the top of the file for easy updating.
+
+See this comment for design rationale:
+https://github.com/pytorch/vision/pull/1321#issuecomment-531033978
+"""
 
 import jinja2
+import yaml
 import os.path
 
-d = os.path.dirname(__file__)
-env = jinja2.Environment(
-    loader=jinja2.FileSystemLoader(d),
-    lstrip_blocks=True,
-    autoescape=False,
-)
-with open(os.path.join(d, 'config.yml'), 'w') as f:
-    f.write(env.get_template('config.yml.in').render())
+
+def workflows(prefix='', upload=False, filter_branch=None, indentation=6):
+    w = []
+    for btype in ["wheel", "conda"]:
+        for os_type in ["linux", "macos"]:
+            for python_version in ["2.7", "3.5", "3.6", "3.7"]:
+                for unicode in ([False, True] if btype == "wheel" and python_version == "2.7" else [False]):
+                    w += workflow_pair(btype, os_type, python_version, unicode, filter_branch, prefix, upload)
+
+    return indent(indentation, w)
+
+
+def workflow_pair(btype, os_type, python_version, unicode, filter_branch, prefix='', upload=False):
+
+    w = []
+    unicode_suffix = "_unicode" if unicode else ""
+    base_workflow_name = f"{prefix}binary_{os_type}_{btype}_py{python_version}{unicode_suffix}"
+
+    w.append(generate_base_workflow(base_workflow_name, python_version, unicode, filter_branch, os_type, btype))
+
+    if upload:
+
+        is_py3_linux = os_type == 'linux' and not python_version.startswith("2.")
+
+        # XXX This logic is suspect...
+        upload_job_filter_branch = not is_py3_linux and filter_branch
```

Comment:
Not sure if this was a bug introduced at some point in the in the existing logic.  Currently, the `filters:` yaml chunk is excluded from `*_upload` jobs on Linux with Python 3, so I've crafted my logic to match this behavior.  Not sure what the rationale is, though, or whether that's intentional.

Reply:
Is this code generation (with exception) the same across domains? What are the differences?

Reply:
There are some minor differences with the `vision` repo's `config.yml` file, and by extension the codegen.  To name a few:
* `vision` exercises an extra dimension of configuration with different versions of CUDA + cpu
* it does not have `smoke_test` jobs
* it lacks `filters:` sections in the YAML